controllers/cluster/capabilities.go

```diff
@@ -0,0 +1,168 @@
+		}
+	}
+
+	return Capabilities{
+		IsOnOpenshift: onOpenshift,
+		HasMachineAPI: machineAPI,
+	}, nil
+}
+
+// isOnOpenshift returns true if the cluster has the openshift config group
+func isOnOpenshift(config *rest.Config) (bool, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return false, err
+	}
+	apiGroups, err := dc.ServerGroups()
+	kind := schema.GroupVersionKind{Group: "config.openshift.io", Version: "v1", Kind: "ClusterVersion"}
+	for _, apiGroup := range apiGroups.Groups {
+		for _, supportedVersion := range apiGroup.Versions {
+			if supportedVersion.GroupVersion == kind.GroupVersion().String() {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// hasMachineAPI returns true it the cluster has the MachineAPI
+func hasMachineAPI(config *rest.Config) (bool, error) {
+	client, err := configv1.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create cluster version client")
+	}
+	ocpVersion, err := getOpenshiftVersion(client)
+	if err != nil {
+		return false, errors.Wrap(err, "could not get OCP version")
+	}
+
+	if ocpVersion.Major == 4 && ocpVersion.Minor >= 14 {
+		return hasMachineAPICapability(client)
+	} else {
+		return hasMachineAPIGroup(config)
+	}
+}
+
+// getOpenshiftVersion returns the OpenShift version of the cluster
+func getOpenshiftVersion(client *configv1.ConfigV1Client) (*ServerVersion, error) {
+	clusterVersion, err := client.ClusterVersions().Get(context.TODO(), "version", metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get OCP version")
+	}
+
+	var ocpVersion string
+	for _, update := range clusterVersion.Status.History {
+		if update.State == v1.CompletedUpdate {
+			// obtain the version from the last completed update
+			ocpVersion = update.Version
+			break
+		}
+	}
+	if ocpVersion == "" {
+		return nil, errors.New("could not find up to date OCP version")
+	}
+
+	matches := versionRegex.FindStringSubmatch(ocpVersion)
+	if matches == nil || len(matches) != 3 {
+		return nil, fmt.Errorf("could not get OCP version from %s", ocpVersion)
+	}
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get major version")
+	}
+	minor, err := strconv.Atoi(matches[2])
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get minor version")
+	}
+
+	return &ServerVersion{
+		OcpVersion: ocpVersion,
+		Major:      major,
+		Minor:      minor,
+	}, nil
+}
+
+// hasMachineAPICapability returns true if the cluster has the MachineAPI capability enabled
+func hasMachineAPICapability(c *configv1.ConfigV1Client) (bool, error) {
+	cvs, err := c.ClusterVersions().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get ClusterVersion")
+	}
+
+	for _, cv := range cvs.Items {
+		if cv.Status.Capabilities.EnabledCapabilities == nil {
```

Comment:
is there guarantee that  `cv.Status.Capabilities` isn't nil ?

Reply:
not sure actually, better verify this case

Reply:
> (and not optional)

But it's also not marked with  **required** so I'm not sure what's the default choice is.
Even though it's hardly a solid proof , according to AI the default is optional - see below.
From a quick search I couldn't find one way or the other in official k8s docs - maybe you can do better :sweat_smile: .

To sum it up I think that it's fine to assume it's required if we can find some strong evidence for that, otherwise I prefer the "better safe than sorry" approach. 

```
So in the context of your question, both History and Capabilities fields are optional, as History is explicitly marked as optional with the // +optional comment annotation, and Capabilities is assumed to be optional by default unless explicitly marked as required.
```

Reply:
It does not matter if this marked as optional or required from an API point of view. When it's not a pointer, it can never be nil. That's just Go. When the field isn't set, it defaults to its zero value, which is an empty ClusterVersionCapabilitiesStatus struct in this case.

Reply:
the more interesting part is `EnabledCapabilities`. It's a slice, which can be nil OR empty. But `for` handles both, so this check is not needed at all :)